src/main/java/org/junit/runners/BlockJUnit4ClassRunner.java

```diff
@@ -380,29 +382,22 @@ protected Statement withAfters(FrameworkMethod method, Object target,
                 target);
     }
 
-    private Statement withRules(FrameworkMethod method, Object target,
-            Statement statement) {
-        List<TestRule> testRules = getTestRules(target);
-        Statement result = statement;
-        result = withMethodRules(method, testRules, target, result);
-        result = withTestRules(method, testRules, result);
-
-        return result;
-    }
-
-    private Statement withMethodRules(FrameworkMethod method, List<TestRule> testRules,
-            Object target, Statement result) {
-        Statement withMethodRules = result;
-        for (org.junit.rules.MethodRule each : getMethodRules(target)) {
-            if (!(each instanceof TestRule && testRules.contains(each))) {
-                withMethodRules = each.apply(withMethodRules, method, target);
+    private Statement withRules(FrameworkMethod method, Object target, Statement statement) {
+        RuleContainer ruleContainer = new RuleContainer();
+        RuleContainer.CURRENT.set(ruleContainer);
+        try {
+            List<TestRule> testRules = getTestRules(target);
+            List<MethodRule> methodRules = rules(target);
```

Comment:
Since `methodRules` is used once, I suggest inlining it:

    for (MethodRule each : rules(target)) {
      ...

(no idea why MethodRule was fully qualified here)